File: python/relationship_store.py
# =============================================================================
# vapor-idx — relationship_store.py
# =============================================================================

from __future__ import annotations
import time
from collections import defaultdict
from .types import VaporRelationship, VaporSchema, VaporError, VaporSchemaError


class RelationshipStore:
    def __init__(self, schema: VaporSchema) -> None:
        self._schema  = schema
        self._edges:  dict[str, VaporRelationship] = {}
        self._out:    dict[str, dict[str, set[str]]] = defaultdict(lambda: defaultdict(set))
        self._inc:    dict[str, dict[str, set[str]]] = defaultdict(lambda: defaultdict(set))
        self._by_type: dict[str, set[str]] = defaultdict(set)
        self._counter = 0

    def relate(
        self,
        source_id: str, source_type: str,
        rel_type: str,
        target_id: str, target_type: str,
        metadata: dict,
    ) -> str:
        type_def = self._schema.types.get(source_type)
        if type_def is None:
            raise VaporError(f'Unknown source type "{source_type}".')
        rel_def = type_def.relationships.get(rel_type)
        if rel_def is None:
            raise VaporSchemaError(
                f'Relationship "{rel_type}" not declared on type "{source_type}".'
            )
        if "*" not in rel_def.targetTypes and target_type not in rel_def.targetTypes:
            raise VaporSchemaError(
                f'Relationship "{rel_type}" from "{source_type}" does not allow '
                f'target type "{target_type}". Allowed: {list(rel_def.targetTypes)}.'
            )
        self._enforce_cardinality(source_id, target_id, rel_type, rel_def.cardinality)

        edge_id = self._next_id()
        now     = int(time.time() * 1000)
        edge    = VaporRelationship(
            id=edge_id, relationship_type=rel_type,
            source_id=source_id, target_id=target_id,
            metadata=dict(metadata), _created_at=now,
        )
        self._edges[edge_id] = edge
        self._add_adjacency(source_id, target_id, rel_type, edge_id)
        self._by_type[rel_type].add(edge_id)

        if not rel_def.directed:
            rev_id  = self._next_id()
            rev_meta = {**metadata, "_reverse": True}
            rev_edge = VaporRelationship(
                id=rev_id, relationship_type=rel_type,
                source_id=target_id, target_id=source_id,
                metadata=rev_meta, _created_at=now,
            )
            self._edges[rev_id] = rev_edge
            self._add_adjacency(target_id, source_id, rel_type, rev_id)
            self._by_type[rel_type].add(rev_id)

        return edge_id

    def unrelate(self, edge_id: str) -> None:
        edge = self._edges.pop(edge_id, None)
        if edge is None:
            return
        self._out[edge.source_id][edge.relationship_type].discard(edge_id)
        self._inc[edge.target_id][edge.relationship_type].discard(edge_id)
        self._by_type[edge.relationship_type].discard(edge_id)

    def remove_for_record(self, record_id: str) -> None:
        to_remove: list[str] = []
        for edgeset in self._out.get(record_id, {}).values():
            to_remove.extend(edgeset)
        for edgeset in self._inc.get(record_id, {}).values():
            to_remove.extend(edgeset)
        for eid in to_remove:
            self.unrelate(eid)

    def get_edges_for_record(
        self,
        record_id: str,
        rel_type: str | None = None,
        direction: str = "both",
    ) -> list[VaporRelationship]:
        edge_ids: set[str] = set()
        if direction in ("outgoing", "both"):
            out = self._out.get(record_id, {})
            if rel_type:
                edge_ids.update(out.get(rel_type, set()))
            else:
                for s in out.values():
                    edge_ids.update(s)
        if direction in ("incoming", "both"):
            inc = self._inc.get(record_id, {})
            if rel_type:
                edge_ids.update(inc.get(rel_type, set()))
            else:
                for s in inc.values():
                    edge_ids.update(s)
        return [self._edges[eid] for eid in edge_ids if eid in self._edges]

    def get_neighbour_ids(self, record_id: str, rel_type: str, direction: str = "outgoing") -> list[str]:
        edges = self.get_edges_for_record(record_id, rel_type, direction)
        return [e.target_id if e.source_id == record_id else e.source_id for e in edges]

    def get_all(self) -> list[VaporRelationship]:
        return list(self._edges.values())

    @property
    def total_edges(self) -> int:
        return len(self._edges)

    @property
    def edge_counts_by_type(self) -> dict[str, int]:
        return {t: len(s) for t, s in self._by_type.items()}

    def clear(self) -> None:
        self._edges.clear()
        self._out.clear()
        self._inc.clear()
        self._by_type.clear()

    def _enforce_cardinality(self, source_id, target_id, rel_type, cardinality) -> None:
        if cardinality == "many-to-many":
            return
        out_count = len(self._out.get(source_id, {}).get(rel_type, set()))
        if cardinality in ("one-to-one", "one-to-many") and out_count > 0:
            if cardinality == "one-to-one":
                raise VaporError(f'Cardinality violation: "{rel_type}" is one-to-one.')
        if cardinality == "one-to-one":
            in_count = len(self._inc.get(target_id, {}).get(rel_type, set()))
            if in_count > 0:
                raise VaporError(f'Cardinality violation: "{rel_type}" target already has incoming edge.')

    def _add_adjacency(self, source_id, target_id, rel_type, edge_id) -> None:
        self._out[source_id][rel_type].add(edge_id)
        self._inc[target_id][rel_type].add(edge_id)

    def _next_id(self) -> str:
        self._counter += 1
        return f"vpe_{int(time.time()*1000):x}_{self._counter:x}"

File: rust/src/relationship_store.rs
// =============================================================================
// vapor-idx — relationship_store.rs
// First-class relationship edge storage with bidirectional adjacency.
// FIX: extract `directed` from rel_def BEFORE calling self.next_id() (mutable)
//      to avoid the Rust borrow checker rejecting simultaneous immutable +
//      mutable borrows of self.
// =============================================================================

use crate::types::*;
use std::collections::{HashMap, HashSet};
use std::time::{SystemTime, UNIX_EPOCH};

pub struct RelationshipStore {
    schema: VaporSchema,
    edges: HashMap<String, VaporRelationship>,
    outgoing: HashMap<String, HashMap<String, HashSet<String>>>,
    incoming: HashMap<String, HashMap<String, HashSet<String>>>,
    by_type: HashMap<String, HashSet<String>>,
    counter: u64,
}

impl RelationshipStore {
    pub fn new(schema: VaporSchema) -> Self {
        Self {
            schema,
            edges: HashMap::new(),
            outgoing: HashMap::new(),
            incoming: HashMap::new(),
            by_type: HashMap::new(),
            counter: 0,
        }
    }

    pub fn relate(
        &mut self,
        source_id: &str,
        source_type: &str,
        rel_type: &str,
        target_id: &str,
        target_type: &str,
        metadata: serde_json::Value,
    ) -> VaporResult<String> {
        let type_def = self
            .schema
            .types
            .get(source_type)
            .ok_or_else(|| VaporError::NotFound(format!("Type \"{source_type}\" not found")))?;

        let rel_def = type_def.relationships.get(rel_type).ok_or_else(|| {
            VaporError::Schema(format!(
                "Relationship \"{rel_type}\" not declared on type \"{source_type}\"."
            ))
        })?;

        if !rel_def
            .target_types
            .iter()
            .any(|t| t == "*" || t == target_type)
        {
            return Err(VaporError::Schema(format!(
                "Relationship \"{rel_type}\" from \"{source_type}\" does not allow \
                 target \"{target_type}\"."
            )));
        }

        self.enforce_cardinality(source_id, target_id, rel_type, &rel_def.cardinality)?;

        // Extract `directed` BEFORE any mutable borrow of self.
        // Rust cannot hold &self.schema (via rel_def) and call &mut self methods
        // simultaneously. Copying the bool here ends the immutable borrow.
        let directed = rel_def.directed;

        let edge_id = self.next_id();
        let now = now_ms();

        let edge = VaporRelationship {
            id: edge_id.clone(),
            relationship_type: rel_type.to_owned(),
            source_id: source_id.to_owned(),
            target_id: target_id.to_owned(),
            metadata: metadata.clone(),
            created_at: now,
        };

        self.edges.insert(edge_id.clone(), edge);
        self.add_adjacency(source_id, target_id, rel_type, &edge_id);
        self.by_type
            .entry(rel_type.to_owned())
            .or_default()
            .insert(edge_id.clone());

        if !directed {
            let rev_id = self.next_id();
            let rev_meta = {
                let mut m = metadata.as_object().cloned().unwrap_or_default();
                m.insert("_reverse".into(), serde_json::Value::Bool(true));
                serde_json::Value::Object(m)
            };
            let rev_edge = VaporRelationship {
                id: rev_id.clone(),
                relationship_type: rel_type.to_owned(),
                source_id: target_id.to_owned(),
                target_id: source_id.to_owned(),
                metadata: rev_meta,
                created_at: now,
            };
            self.edges.insert(rev_id.clone(), rev_edge);
            self.add_adjacency(target_id, source_id, rel_type, &rev_id);
            self.by_type
                .entry(rel_type.to_owned())
                .or_default()
                .insert(rev_id);
        }

        Ok(edge_id)
    }

    pub fn unrelate(&mut self, edge_id: &str) {
        if let Some(edge) = self.edges.remove(edge_id) {
            self.outgoing
                .get_mut(&edge.source_id)
                .and_then(|m| m.get_mut(&edge.relationship_type))
                .map(|s| s.remove(edge_id));
            self.incoming
                .get_mut(&edge.target_id)
                .and_then(|m| m.get_mut(&edge.relationship_type))
                .map(|s| s.remove(edge_id));
            self.by_type
                .get_mut(&edge.relationship_type)
                .map(|s| s.remove(edge_id));
        }
    }

    pub fn remove_for_record(&mut self, record_id: &str) {
        let mut to_remove: Vec<String> = Vec::new();
        if let Some(out_map) = self.outgoing.get(record_id) {
            for ids in out_map.values() {
                to_remove.extend(ids.iter().cloned());
            }
        }
        if let Some(in_map) = self.incoming.get(record_id) {
            for ids in in_map.values() {
                to_remove.extend(ids.iter().cloned());
            }
        }
        let unique: Vec<String> = to_remove
            .into_iter()
            .collect::<HashSet<_>>()
            .into_iter()
            .collect();
        for eid in unique {
            self.unrelate(&eid);
        }
    }

    pub fn get_edges_for_record(
        &self,
        record_id: &str,
        rel_type: Option<&str>,
        direction: &str,
    ) -> Vec<VaporRelationship> {
        let mut ids: HashSet<String> = HashSet::new();
        if direction == "outgoing" || direction == "both" {
            if let Some(out) = self.outgoing.get(record_id) {
                match rel_type {
                    Some(t) => {
                        out.get(t).map(|s| ids.extend(s.iter().cloned()));
                    }
                    None => {
                        for s in out.values() {
                            ids.extend(s.iter().cloned());
                        }
                    }
                }
            }
        }
        if direction == "incoming" || direction == "both" {
            if let Some(inc) = self.incoming.get(record_id) {
                match rel_type {
                    Some(t) => {
                        inc.get(t).map(|s| ids.extend(s.iter().cloned()));
                    }
                    None => {
                        for s in inc.values() {
                            ids.extend(s.iter().cloned());
                        }
                    }
                }
            }
        }
        ids.iter()
            .filter_map(|id| self.edges.get(id))
            .cloned()
            .collect()
    }

    pub fn get_neighbour_ids(
        &self,
        record_id: &str,
        rel_type: &str,
        direction: &str,
    ) -> Vec<String> {
        self.get_edges_for_record(record_id, Some(rel_type), direction)
            .into_iter()
            .map(|e| {
                if e.source_id == record_id {
                    e.target_id
                } else {
                    e.source_id
                }
            })
            .collect()
    }

    pub fn get_all(&self) -> Vec<VaporRelationship> {
        self.edges.values().cloned().collect()
    }

    pub fn total_edges(&self) -> usize {
        self.edges.len()
    }

    pub fn edges_by_type(&self) -> HashMap<String, usize> {
        self.by_type
            .iter()
            .map(|(k, v)| (k.clone(), v.len()))
            .collect()
    }

    pub fn clear(&mut self) {
        self.edges.clear();
        self.outgoing.clear();
        self.incoming.clear();
        self.by_type.clear();
    }

    fn enforce_cardinality(
        &self,
        source_id: &str,
        target_id: &str,
        rel_type: &str,
        cardinality: &Cardinality,
    ) -> VaporResult<()> {
        if *cardinality == Cardinality::ManyToMany {
            return Ok(());
        }
        let out_count = self
            .outgoing
            .get(source_id)
            .and_then(|m| m.get(rel_type))
            .map(|s| s.len())
            .unwrap_or(0);
        if *cardinality == Cardinality::OneToOne && out_count > 0 {
            return Err(VaporError::Cardinality(format!(
                "\"{rel_type}\" is one-to-one but source already has outgoing edge."
            )));
        }
        if *cardinality == Cardinality::OneToOne {
            let in_count = self
                .incoming
                .get(target_id)
                .and_then(|m| m.get(rel_type))
                .map(|s| s.len())
                .unwrap_or(0);
            if in_count > 0 {
                return Err(VaporError::Cardinality(format!(
                    "\"{rel_type}\" is one-to-one but target already has incoming edge."
                )));
            }
        }
        Ok(())
    }

    fn add_adjacency(&mut self, source: &str, target: &str, rel_type: &str, edge_id: &str) {
        self.outgoing
            .entry(source.to_owned())
            .or_default()
            .entry(rel_type.to_owned())
            .or_default()
            .insert(edge_id.to_owned());
        self.incoming
            .entry(target.to_owned())
            .or_default()
            .entry(rel_type.to_owned())
            .or_default()
            .insert(edge_id.to_owned());
    }

    fn next_id(&mut self) -> String {
        self.counter += 1;
        format!("vpe_{:x}_{:x}", now_ms(), self.counter)
    }
}

fn now_ms() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_millis() as i64)
        .unwrap_or(0)
}
